src/display/annotation_layer.js

```diff
@@ -1832,157 +1844,216 @@ class ChoiceWidgetAnnotationElement extends WidgetAnnotationElement {
-    // annotation, we cannot hide the entire container as the image would
-    // disappear too. In that special case, hiding the wrapper suffices.
-    this.hideElement = this.hideWrapper ? wrapper : this.container;
-    this.hideElement.hidden = true;
+  #pinned = false;
 
-    const popup = document.createElement("div");
-    popup.classList.add("popup");
+  #popup = null;
 
-    const color = this.color;
-    if (color) {
-      // Enlighten the color.
-      const r = BACKGROUND_ENLIGHT * (255 - color[0]) + color[0];
-      const g = BACKGROUND_ENLIGHT * (255 - color[1]) + color[1];
-      const b = BACKGROUND_ENLIGHT * (255 - color[2]) + color[2];
-      popup.style.backgroundColor = Util.makeHexColor(r | 0, g | 0, b | 0);
+  #rect = null;
+
+  #richText = null;
+
+  #titleObj = null;
+
+  constructor({
+    container,
+    color,
+    elements,
+    titleObj,
+    modificationDate,
+    contentsObj,
+    richText,
+    parent,
+    rect,
+    parentRect,
+    open,
+  }) {
+    this.#container = container;
+    this.#titleObj = titleObj;
+    this.#contentsObj = contentsObj;
+    this.#richText = richText;
+    this.#parent = parent;
+    this.#color = color;
+    this.#rect = rect;
+    this.#parentRect = parentRect;
+    this.#elements = elements;
+
+    const dateObject = PDFDateString.toDateObject(modificationDate);
+    if (dateObject) {
+      // The modification date is shown in the popup instead of the creation
+      // date if it is available and can be parsed correctly, which is
+      // consistent with other viewers such as Adobe Acrobat.
+      this.#dateTimePromise = parent.l10n.get("annotation_date_string", {
+        date: dateObject.toLocaleDateString(),
+        time: dateObject.toLocaleTimeString(),
+      });
+    }
+
+    this.trigger = elements.flatMap(e => e.getElementsToTriggerPopup());
+    // Attach the event listeners to the trigger element.
+    for (const element of this.trigger) {
+      element.addEventListener("click", this.#boundToggle);
+      element.addEventListener("mouseenter", this.#boundShow);
+      element.addEventListener("mouseleave", this.#boundHide);
+      if (typeof PDFJSDev !== "undefined" && PDFJSDev.test("TESTING")) {
+        element.classList.add("popupTriggerArea");
+      }
+    }
+
+    this.#container.hidden = true;
+    if (open) {
+      this.#toggle();
+    }
+  }
+
+  static create(parameters) {
+    return new PopupElement(parameters);
+  }
+
+  render() {
+    if (this.#popup) {
+      return;
     }
 
+    const {
+      page: { view },
+      viewport: {
+        rawDims: { pageWidth, pageHeight, pageX, pageY },
+      },
+    } = this.#parent;
+    const popup = (this.#popup = document.createElement("div"));
+    popup.className = "popup";
+
+    if (this.#color) {
+      const baseColor = (popup.style.outlineColor = Util.makeHexColor(
+        ...this.#color
+      ));
+      popup.style.backgroundColor = `color-mix(in srgb, ${baseColor} 30%, white)`;
```

Comment:
This unfortunately looks too new to be generally usable, given https://developer.mozilla.org/en-US/docs/Web/CSS/color_value/color-mix#browser_compatibility

Reply:
Having said that, what does the fallback behaviour look like in browsers without support for this?

If the Popup is still visible/usable, and that we ignore the custom colour (as specified in the Annotation), then I'm fine with this as-is :-)